linkerd/app/inbound/src/http/router.rs

```diff
@@ -415,3 +423,35 @@ impl Param<transport::labels::Key> for Http {
         transport::labels::Key::InboundClient
     }
 }
+
+// === impl ClientRescue ===
+
+impl ClientRescue {
+    pub fn layer() -> errors::respond::Layer<Self> {
+        errors::respond::NewRespond::layer(Self)
+    }
+}
+
+impl errors::HttpRescue<Error> for ClientRescue {
+    fn rescue(&self, error: Error) -> Result<errors::SyntheticHttpResponse> {
+        if Self::has_cause::<std::io::Error>(&*error) {
+            return Ok(errors::SyntheticHttpResponse {
+                http_status: http::StatusCode::BAD_GATEWAY,
+                grpc_status: errors::Grpc::Unavailable,
+                close_connection: true,
+                message: error.to_string(),
+            });
+        }
+
+        if Self::has_cause::<errors::ConnectTimeout>(&*error) {
+            return Ok(errors::SyntheticHttpResponse {
+                http_status: http::StatusCode::GATEWAY_TIMEOUT,
+                grpc_status: errors::Grpc::DeadlineExceeded,
+                close_connection: true,
+                message: error.to_string(),
+            });
+        }
```

Comment:
it's a little bit of a bummer that the two `has_cause` calls will traverse the cause chain twice...ideally, we could do it once, and check for _both_ types at every cause. but, this isn't really possible with the `has_cause` function, since it would have to be generic over a variadic number of type parameters, which isn't possible. i guess the alternative would just be to not use that function.

this is probably fine, though, and it's clearer than rewriting the cause chain traversal here. i don't know if this is hot enough for it to matter?